server/src/main/java/org/jboss/as/server/controller/resources/ModuleInfoHandler.java

```diff
@@ -0,0 +1,209 @@
+            .build();
+
+    public static ModuleInfoHandler INSTANCE = new ModuleInfoHandler();
+
+    protected void populateModel(final ModelNode operation, final ModelNode model) throws OperationFailedException {
+        for (AttributeDefinition attr : DEFINITION.getParameters()) {
+            attr.validateAndSet(operation, model);
+        }
+    }
+
+    @Override
+    public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+        ModelNode model = new ModelNode();
+        populateModel(operation, model);
+        String moduleName = MODULE_NAME.resolveModelAttribute(context, model).asString();
+        String slot = MODULE_SLOT.resolveModelAttribute(context, model).asString();
+        ModuleIdentifier id = ModuleIdentifier.create(moduleName, slot);
+        ModuleLoader loader = Module.getBootModuleLoader();
+        try {
+            ModuleLoaderMXBean mxBean = getMxBean(loader);
+            ModuleInfo moduleInfo = mxBean.getModuleDescription(id.toString());
+            context.getResult().set(displayModuleInfo(moduleInfo));
+        } catch (Exception e) {
+            throw ServerLogger.ROOT_LOGGER.couldNotGetModuleInfo(id.toString(), e);
+        }
+    }
+
+
+    private ModuleLoaderMXBean getMxBean(ModuleLoader loader) throws ReflectiveOperationException {
+        Field mxBeanField = ModuleLoader.class.getDeclaredField("mxBean");
+        mxBeanField.setAccessible(true);
+        return (ModuleLoaderMXBean) mxBeanField.get(loader);
+    }
+
+    /*
+    Here is the information to expose (use ModuleLoaderMXBean as a source of information):
+
+        Loaded module names
+        For each module:
+            Module name
+            Main class name (if any)
+            Class loader name string
+            Fallback loader name string
+            Dependency information:
+                Dependency type
+                Export filter string
+                Import filter string
+                Dependency module name
+                Is-optional flag
+                Local loader type class name
+                Local loader paths
+            Resource loader information for each module:
+                Resource loader type string
+                Resource loader paths
+
+    Note that there are multiple module loaders. For our purposes there should be (at least) two categories displayed:
+
+        The static module loader
+        The deployment module loader
+
+
+     */
+    private ModelNode displayModuleInfo(ModuleInfo module) throws Exception {
+        ModelNode result = new ModelNode();
+        result.get("name").set(module.getName());
+
+        if (module.getMainClass() != null) {
+            result.get("main-class").set(module.getMainClass());
+        }
+        if (module.getFallbackLoader() != null) {
+            result.get("fallback-loader").set(module.getFallbackLoader());
+        }
+
+        ModelNode dependencies = result.get("dependencies");
```

Comment:
This should be ModelNode dependencies = result.get("dependencies").setEmptyList(), or DEPENDENCIES should be configured to allowNull.

I think the former is better.